egressgw: change special values for gatewayIP

MrFreezeex opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/master/USERS.md)
- [ ] Thanks for contributing!

- Change excluded CIDR "special" value of gatewayIP from 0.0.0.0 -> 0.0.0.1

MrFreezeex committed:
egressgw: change special values for gatewayIP

0.0.0.0 now represents that the gateway has not been found (while it
previously meant gateway not found AND excluded CIDR). While 0.0.0.1
now represents an excluded CIDR. Future special values should all be
included in the 0.0.0.0/8 range.

Both values (0.0.0.0, 0.0.0.1) currently behave like excluded CIDR and
skip the egressgw NAT-ing while not dropping the packet for backward
compatibilities during Cilium upgrade. In a future release gateway not
found (0.0.0.0) will drop the traffic.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

ldelossa reviewed (approved):
The change looks reasonable however, I'd like either @julianwiedmann or @jibi to give a final ack for full context.

MrFreezeex committed:
egressgw: add tests case if the gateway is not found

Add a new test to check the case whenever the egress gateway doesn't match
any gateway.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

nebril commented on test/k8s/egress.go:
```diff
@@ -408,6 +408,47 @@ var _ = SkipDescribeIf(func() bool {
 				})
 			})
 		})
+		Context("egress gw policy when the gateway is not found", func() {
```
Too bad we are adding new ginkgo tests, but AFAIU egressgw tests are not yet ported to CI 3.0. I guess it's OK to add new test cases here then.

jibi replied:
:100:

jibi commented on bpf/lib/egress_policies.h:
```diff
@@ -38,11 +38,12 @@ bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint)
 	 * skip redirection
 	 */
-	if (!egress_gw_policy->gateway_ip)
+	if (!egress_gw_policy->gateway_ip || egress_gw_policy->gateway_ip == 1)
```
I think at this point we should define a couple of constants at the top of this header and use them here, something like `EGRESS_GATEWAY_NO_GATEWAY` and `EGRESS_GATEWAY_EXCLUDED_CIDR`

jibi commented on pkg/egressgateway/manager.go:
```diff
@@ -31,8 +31,9 @@ import (
-	zeroIPv4 = net.ParseIP("0.0.0.0")
+	log                                 = logging.DefaultLogger.WithField(logfields.LogSubsys, "egressgateway")
+	notFoundGatewaySpecialValueIPv4     = net.ParseIP("0.0.0.0")
+	excludedCIDRGatewaySpecialValueIPv4 = net.ParseIP("0.0.0.1")
```
We can already export these as we'll need them in the next PR (to translate these values in the `bpf egress list` command):

```suggestion
        // GatewayNotFoundIPv4 is a special IP value used as gatewayIP in the BPF policy map to indicate no gateway was found for the given policy
	GatewayNotFoundIPv4     = net.ParseIP("0.0.0.0")
	
	// ExcludedCIDRIPv4 is a special IP value used as gatewayIP in the BPF policy map to indicate the entry is for an excluded CIDR and should skip egress gateway
	ExcludedCIDRIPv4 = net.ParseIP("0.0.0.1")
```

MrFreezeex replied:
Since you mentioned bpf egress list, I'm already adding the code to translate it :D. I think it should be ok to do it here even though we are not already dropping the traffic.

jibi commented on test/k8s/egress.go:
```diff
@@ -408,6 +408,47 @@ var _ = SkipDescribeIf(func() bool {
+
+				// Wait for 2 entries:
+				// - 1 policies, each with:
+				//   - 2 matching endpoints
```
```suggestion
				// Wait for 2 entries:
				// - 1 policy with 2 matching endpoints
```

jibi reviewed (commented):
Looking good, thanks! Just a couple of nits

MrFreezeex committed:
egressgw: translate gateway IP special values in bpf egress list

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

jibi commented on bpf/lib/egress_policies.h:
```diff
@@ -38,10 +43,9 @@ bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint)
-	if (!egress_gw_policy->gateway_ip)
+	/* FIXME: no gateway traffic should be dropped in a future release */
+	if (egress_gw_policy->gateway_ip == EGRESS_GATEWAY_NO_GATEWAY ||
+            egress_gw_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)
```
BPF checkpatch is complaining about this:
```
   Error: ERROR:CODE_INDENT: code indent should use tabs where possible
  #44: FILE: bpf/lib/egress_policies.h:48:
  +            egress_gw_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)$
  
  CHECK:PARENTHESIS_ALIGNMENT: Alignment should match open parenthesis
  #44: FILE: bpf/lib/egress_policies.h:48:
  +	if (egress_gw_policy->gateway_ip == EGRESS_GATEWAY_NO_GATEWAY ||
  +            egress_gw_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)
```

but let's wait until the tests finish running before pushing again

jibi commented on bpf/lib/egress_policies.h:
```diff
@@ -38,10 +43,9 @@ bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint)
-	 * skip redirection
-	 */
-	if (!egress_gw_policy->gateway_ip)
+	/* FIXME: no gateway traffic should be dropped in a future release */
```
Almost there :slightly_smiling_face:

`K8sDatapathEgressGatewayTest tunnel disabled with endpointRoutes enabled egress gw policy with exclusion CIDRs` is failing as we need to handle this new case also in `egress_gw_snat_needed()` and `egress_gw_reply_needs_redirect()`

MrFreezeex replied:
Indeed I updated those with the new macro now, should be good hopefully :crossed_fingers:

jibi commented on bpf/lib/egress_policies.h:
```diff
@@ -87,10 +88,7 @@ bool egress_gw_reply_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint,
-	 * skip reply redirect
-	 */
-	if (!egress_policy->gateway_ip)
+	if (egress_gw_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)
```
I think we still need to handle the `EGRESS_GATEWAY_NO_GATEWAY` case (ditto for `egress_gw_snat_needed`).

Also `s/egress_gw_policy/egress_policy/`

MrFreezeex replied:
Hmm I'm not sure the no gateway case need to be handled here :thinking:. Those two functions are on the "gateway node" so it means that we would encounter this case whenever loosing the gateway node but traffic still get to it for some reasons. So I don't think it matters that much to snat or not in `egress_gw_reply_needs_redirect` but perhaps we could also drop the traffic in `egress_gw_reply_needs_redirect` whenever this happen (in stage 2 of the change then). WDYT?

And indeed bad copy paste for the `egress_gw_policy` thing :man_facepalming:.

MrFreezeex replied:
Ah and also doing something special here could maybe be racy? Like if go from no gateway -> some gateway is set, and if the agent wherever the pod is scheduled get the information first... It shouldn't be a huge issue though as it will likely go away very soon and if it was previously in a bad state connection was broken anyway at some point so it simply add a very small delay.

MrFreezeex replied:
Well actually you were totally right sorry, I thought this was somehow executed only on the gateway node but it is done also on the "client node" as well. And yeah not handling this mean that traffic will get snated to 0.0.0.0 :sweat:... It should works now. In the last test run there was something about the egress policy entry not being created on the new "no gateway test" in `test/k8s/egress.go` but I cannot reproduce locally :thinking:...

jibi replied:
> Well actually you were totally right sorry, I thought this was somehow executed only on the gateway node but it is done also on the "client node" as well.

yep, we need this for the case where the source pod is running on the same node configured as gateway for that pod :+1:

asauber commented on bpf/tests/tc_egressgw_redirect.c:
```diff
@@ -270,3 +270,127 @@ int egressgw_skip_excluded_cidr_redirect_check(const struct __ctx_buff *ctx)
 	test_finish();
 }
+
+/* Test that a packet matching an egress gateway policy without a gatewat on the
```
```suggestion
/* Test that a packet matching an egress gateway policy without a gateway on the
```

asauber commented on bpf/tests/tc_egressgw_redirect.c:
```diff
@@ -270,3 +270,127 @@ int egressgw_skip_excluded_cidr_redirect_check(const struct __ctx_buff *ctx)
+		.deny = 0,
+	};
+
+	/* avoid policy drop */
```
(very nit:) Capitalization

jibi commented on bpf/lib/egress_policies.h:
```diff
@@ -16,6 +16,11 @@
+ * case for in the egress gateway policies handling.
+ */
+#define EGRESS_GATEWAY_NO_GATEWAY (0)
+#define EGRESS_GATEWAY_EXCLUDED_CIDR (1)
```
I think this needs to be:

```suggestion
#define EGRESS_GATEWAY_EXCLUDED_CIDR bpf_htonl(1)
```

as the gateway IP is encoded as BE in the policy map

MrFreezeex replied:
Oh yes indeed good catch, thanks :).

jibi commented on bpf/tests/tc_egressgw_redirect.c:
```diff
@@ -270,3 +270,127 @@ int egressgw_skip_excluded_cidr_redirect_check(const struct __ctx_buff *ctx)
+		.gateway_ip = NODE_IP,
+	};
+
+	map_update_elem(&EGRESS_POLICY_MAP, &in_key, &in_val, 0);
```
nit: we don't need this entry, just the one with the 0.0.0.0 gateway should enough for this test

MrFreezeex replied:
Sure! I just did that and rebased (as I heard it fixes the eks conformance test :p)

jibi reviewed (approved):
Looks good now, thanks for bearing with me through the few iterations and for all the extra tests!

With regard to part 2: I don't think we need to wait for anything in particular, we can already start getting that into master, and label it for backport only once this PR is backported and released in v1.13
